transport: remove throwing `protocol_exception` on connection start

DarioMirovic opened the pull request:
`protocol_exception` is thrown in several places. This has become a performance issue, especially when starting/restarting a server. To alleviate this issue, throwing the exception has to be replaced with returning it as a result or an exceptional future.

This PR replaces throws in the `transport/server` module. This is achieved by using result_with_exception, and in some places, where suitable, just by creating and returning an exceptional future.

There are four commits in this PR. The first commit introduces tests in `test/cqlpy`. The second commit refactors transport server `handle_error` to not rethrow exceptions. The third commit refactors reusable buffer writer callbacks. The fourth commit replaces throwing `protocol_exception` to returning it.

Based on the comments on an issue linked in https://github.com/scylladb/scylladb/issues/24567, the main culprit from the side of protocol exceptions is the invalid protocol version one, so I tested that exception for performance.

In order to see if there is a measurable difference, a modified version of `test_protocol_version_mismatch` Python is used, with 100'000 runs across 10 processes (not threads, to avoid Python GIL). One test run consisted of 1 warm-up run and 5 measured runs. First test run has been executed on the current code, with throwing protocol exceptions. Second test urn has been executed on the new code, with returning protocol exceptions. The performance report is in https://github.com/scylladb/scylladb/pull/24738#issuecomment-3051611069. It shows ~10% gains in real, user, and sys time for this test.

Testing

Build: `release`

Test file: `test/cqlpy/test_protocol_exceptions.py`
Test name: `test_protocol_version_mismatch` (modified for mass connection requests)

Test arguments:
```
max_attempts=100'000
num_parallel=10
# total runs = 1'000'000
```

Throwing `protocol_exception` results:
```
real=1:26.97  user=10:00.27  sys=2:34.55  cpu=867%
real=1:26.95  user=9:57.10  sys=2:32.50  cpu=862%
real=1:26.93  user=9:56.54  sys=2:35.59  cpu=865%
real=1:26.96  user=9:54.95  sys=2:32.33  cpu=859%
real=1:26.96  user=9:53.39  sys=2:33.58  cpu=859%

real=1:26.95 user=9:56.85 sys=2:34.11 cpu=862%   # average
```

Returning `protocol_exception` as `result_with_exception` or an exceptional future:
```
real=1:18.46  user=9:12.21  sys=2:19.08  cpu=881%
real=1:18.44  user=9:04.03  sys=2:17.91  cpu=869%
real=1:18.47  user=9:12.94  sys=2:19.68  cpu=882%
real=1:18.49  user=9:13.60  sys=2:19.88  cpu=883%
real=1:18.48  user=9:11.76  sys=2:17.32  cpu=878%

real=1:18.47 user=9:10.91 sys=2:18.77 cpu=879%   # average
```

This PR replaced `transport/server` throws of `protocol_exception` with returns. There are a few other places where protocol exceptions are thrown, and there are many places where `invalid_request_exception` is thrown. That is out of scope of this single PR, so the PR just refs, and does not resolve issue #24567.

Refs: #24567

This PR improves performance in cases when protocol exceptions happen, for example during connection storms. It will require backporting.

andrzej-jackowski-scylladb commented on transport/server.cc:
```diff
@@ -355,11 +357,12 @@ cql_server::connection::parse_frame(temporary_buffer<char> buf) const {
     }
     default:
-        throw exceptions::protocol_exception(format("Invalid or unsupported protocol version: {:d}", _version));
+        return exceptions::protocol_exception(
```
I think it is better to not add new newlines while introducing changes like that (there are many changes like this in this commit)

andrzej-jackowski-scylladb commented on transport/server.cc:
```diff
@@ -448,29 +458,29 @@ future<foreign_ptr<std::unique_ptr<cql_server::response>>>
     auto linearization_buffer = std::make_unique<bytes_ostream>();
     auto linearization_buffer_ptr = linearization_buffer.get();
     return futurize_invoke([this, cqlop, stream, &fbuf, &client_state, linearization_buffer_ptr, permit = std::move(permit), trace_state] () mutable {
+        tracing::set_username(trace_state, client_state.user());
```
Why set_username is moved? If it is really necessary, this change should be rather in a separate commit with a detailed description.

andrzej-jackowski-scylladb commented on transport/server.cc:
```diff
@@ -852,27 +862,30 @@ future<fragmented_temporary_buffer> cql_server::connection::read_and_decompress_
                 }
                 auto in = input_buffer.get_linearized_view(v);
-                return output_buffer.make_fragmented_temporary_buffer(uncomp_len, [&in] (bytes_mutable_view out) {
+                auto buf_res = output_buffer.make_fragmented_temporary_buffer(uncomp_len, [&in] (bytes_mutable_view out) -> utils::result_with_exception<size_t, std::runtime_error> {
```
I'd rather just `return utils::result_into_future(...`

avikivity commented on test/boost/reusable_buffer_test.cc:
```diff
@@ -75,26 +75,34 @@ SEASTAR_TEST_CASE(test_make_buffer) {
 
+        using make_buffer_return_type = utils::result_with_exception<
+            size_t,
+            std::exception
```
I guess it doesn't trigger because you never return an exception, but then, why make the change in the first place?

avikivity replied:
And why not code the actual exception there?

DarioMirovic replied:
@avikivity , what @piodul said is correct. We now return a result from the lambda, and we can not have empty `utils::exception_container`. This test lambda does not return any exception.

I have not coded the actual exception inside the lambda because it would just delay the assert location from inside the lambda to outside, after the lambda call.

Options that come to my mind:
- use `std::exception` in the type definition
    - the exception is never used
- a dummy exception that inherits `std::exception`
    - not really different from the first solution
- allow empty exception container or `resultize` a value as @piodul suggested
    - mixing infrastructure with the fix, maybe create another PR?
- change the lambda in the test to not use asserts, but perhaps return `std::runtime_error` or some other exception
    - delaying asserts; also the exception would not be handled differently - just the assert would be after the lambda call, not inside it

avikivity commented on transport/server.cc:
```diff
@@ -403,7 +409,11 @@ cql_server::connection::read_frame() {
-            return make_ready_future<ret_type>(parse_frame(std::move(buf)));
+            auto frame_res = parse_frame(std::move(buf));
+            if (!frame_res) {
+                return std::move(frame_res).assume_error().as_exception_future<ret_type>();
```
It is better (but more work) to keep returning outcomes rather than exceptions.

avikivity commented on utils/reusable_buffer.hh:
```diff
@@ -9,13 +9,27 @@
+
+template<typename Fn>
+concept fragmented_buffer_writer = buffer_writer_for<Fn, fragmented_temporary_buffer>;
+
```
Why these changes?

DarioMirovic replied:
`reusable_buffer` needs functions with specific signature. While these concepts are not technically required, and what they achieve could be achieved in another way, they do make the compiler fail early and with clearer error messages in case an incompatible function is passed.

DarioMirovic replied:
The patch does not introduce `reusable_buffer`. It is already used inside `transport/server.cc`.

This patch just changes the lambda signature from returning a value to returning result with either a value or an exception. This is done because we don't want to throw, and the previous lambda version that is passed to the `reusable_buffer` methods either returned a value or threw an exception.

DarioMirovic commented:
I have run  `test_protocol_version_mismatch` test from `test/cqlpy/test_protocol_exceptions.py` file on my computer.

Built: `release`

Test arguments:
```
max_attempts=100'000
num_parallel=10
# total runs - 1'000'000
```

Throwing `protocol_exception` results:
```
real=4:57.03  user=4:41.20  sys=1:15.37  cpu=120%
real=4:54.98  user=4:39.63  sys=1:15.29  cpu=120%
real=4:52.88  user=4:37.49  sys=1:14.42  cpu=120%
real=4:52.85  user=4:37.80  sys=1:14.62  cpu=120%
real=4:52.85  user=4:37.96  sys=1:15.23  cpu=120%

real=4:54.12  user=4:38.82  sys=1:14.99  cpu=120%   # average
```

Returning `protocol_exception` as `result_with_exception` or an exceptional future:
```
real=4:50.76  user=4:33.94  sys=1:14.68  cpu=119%
real=4:50.83  user=4:34.27  sys=1:14.60  cpu=119%
real=4:50.78  user=4:34.31  sys=1:14.50  cpu=119%
real=4:50.72  user=4:33.94  sys=1:13.30  cpu=119%
real=4:52.83  user=4:35.58  sys=1:14.19  cpu=119%

real=4:51.18  user=4:34.41  sys=1:14.25  cpu=119%   # average
```

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

| Mode | Stage | Status | Node |
|------|-------|--------|------|
|  | [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18117/) | :white_check_mark: | thor |
| dev | [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18117/) | :white_check_mark: | spider5.cloudius-systems.com |
|  | [test.py](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18117/) | :white_check_mark: | spider5.cloudius-systems.com |
|  | [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18117/)<br> :small_blue_diamond: test/boost/reusable_buffer_test.cc | :white_check_mark: | spider5.cloudius-systems.com |
| release | [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18117/) | :white_check_mark: | spider6.cloudius-systems.com |
|  | [test.py](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18117/) | :white_check_mark: | spider6.cloudius-systems.com |
|  | [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18117/)<br> :small_blue_diamond: test/boost/reusable_buffer_test.cc | :white_check_mark: | spider6.cloudius-systems.com |
| debug | [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18117/) | :white_check_mark: | spider1.cloudius-systems.com |
|  | [test.py](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18117/) | :white_check_mark: | spider1.cloudius-systems.com |
|  | [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18117/)<br> :small_blue_diamond: test/boost/reusable_buffer_test.cc | :white_check_mark: | spider1.cloudius-systems.com |
| dtest | [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/5897/) | :white_check_mark: |  |
|  | [dtest with consistent topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/9124/) | :white_check_mark: |  |
|  | [dtest with gossip topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/17217/) | :white_check_mark: |  |

#### Build Details:
- Duration: 1 hr 16 min

andrzej-jackowski-scylladb commented on test/cqlpy/test_protocol_exceptions.py:
```diff
@@ -0,0 +1,44 @@
+        with cql_bad_protocol(request, host) as session:
+            return 1 if session else 0
+        
+    with concurrent.futures.ThreadPoolExecutor(max_workers=num_parallel) as executor:
```
Is there any difference in `ThreadPoolExecutor` vs `ProcessPoolExecutor` execution time?
With `num_parallel=10`, I imagine, that GIL might be 100% occupied all the time.
I would try with `ProcessPoolExecutor` and if there is a large difference (especially in comparing throwing vs non-protocol_exception case) I would change it (and if not, I will rather keep `ThreadPoolExecutor` as having processes is always an additional burden)

DarioMirovic replied:
It did make significant impact! I'll post the new results below in a new comment.

DarioMirovic commented:
In V3:
- Updated the test to use `ProcessPoolExecutor` instead of `ThreadPoolExecutor`, as per [this comment](https://github.com/scylladb/scylladb/pull/24738#discussion_r2184497119)

DarioMirovic commented:
@avikivity could you take a look, I have made some changes

ptrsmrn commented on test/cqlpy/test_protocol_exceptions.py:
```diff
@@ -0,0 +1,53 @@
+
+    return count
+
+def test_protocol_version_mismatch(scylla_only, request, host):
```
So as discussed, this test needs to be reworked into a functional test, which checks the asserted behaviour just once. Also, if it's not too time consuming, we'd like to have more functional tests that cover other changed code paths.

DarioMirovic replied:
V4 description here: https://github.com/scylladb/scylladb/pull/24738#issuecomment-3069122351

DarioMirovic commented:
In V4:
- Refactored the test with wrong protocol version, named `test_protocol_version_mismatch`, into a functional test
    - Now it executes once, and generates 2 protocol exceptions
- Added three tests that have the same implementation
    - `test_invalid_kind_in_batch_message`
    - `test_unexpected_message_during_auth`
    - `test_no_protocol_exceptions`
        - This one, as the previous two, relies on the same implementation `_protocol_error_impl`, but does not flag any exceptions to be raised inside the test
- The core code is not changed in this revision, just the tests

@avikivity @ptrsmrn

avikivity commented:
> Can you report exceptions thrown per connection before/after? There's a metric for that.

What's the answer to this question?

DarioMirovic commented:
> > Can you report exceptions thrown per connection before/after? There's a metric for that.
> 
> What's the answer to this question?

@avikivity each connection request with bad protocol generates the same invalid protocol protocol exception 2 times. `test_protocol_version_mismatch` expects `metrics_after` to be `metrics_before + 2`, but I have not asserted that way to avoid any potential counter race conditions reading even bigger difference.

nyh reviewed (changes requested):
Thanks. I think the bulk of the patch - changing exceptions to returns - is fine, but there are so many of them that I'm not sure I followed everything. But I guess the compiler would have caught most errors anyway.
I left a few requests in the periphery, such as the tests and utility functions and other stuff.

nyh commented:
> > > Refs: #24567
> > 
> > 
> > Why does it "Refs" and not "Fixes"? Please explain what you didn't solve, and be prepared to also go to that issue, and explain there what you already solved in this pull request, and what still remains to be solved.
> 
> Some `protocol_exception`s have been removed by this PR, but the issue is a bit vague, stating that some `invalid_request_exception`s should also be removed. Since this PR does a subset of the work needed, it refs, and not solves the issue.

@ptrsmrn you opened that issue so I guess for you it's not vague, so can you please consider whether you would consider your issue will be "fixed" by this patch or not, or whether the issue should be kept open with a note that one part of it was done, but this second part still needs to be done?

ptrsmrn commented:
> > > > Refs: #24567
> > > 
> > > 
> > > Why does it "Refs" and not "Fixes"? Please explain what you didn't solve, and be prepared to also go to that issue, and explain there what you already solved in this pull request, and what still remains to be solved.
> > 
> > 
> > Some `protocol_exception`s have been removed by this PR, but the issue is a bit vague, stating that some `invalid_request_exception`s should also be removed. Since this PR does a subset of the work needed, it refs, and not solves the issue.
> 
> @ptrsmrn you opened that issue so I guess for you it's not vague, so can you please consider whether you would consider your issue will be "fixed" by this patch or not, or whether the issue should be kept open with a note that one part of it was done, but this second part still needs to be done?

@nyh the referenced issue is opened against fixing both the protocol_exception and invalid_request_exception. This PR doesn't resolve all the protocol_exception cases, a follow-up is needed. 
We could split the issue into 2 to keep each exception separate and the follow up PR would fix (close) one of them.

nuivall commented on test/cqlpy/test_protocol_exceptions.py:
```diff
@@ -0,0 +1,156 @@
+            yield session
+            session.shutdown()
+    except NoHostAvailable:
+        yield None
```
I am not sure why this indirection is even needed. Afaics it only check if cql_session(...protocol_version=5) returns an exception

DarioMirovic replied:
@nyh I have added a successful connection step in the test, with protocol version 4, before proceeding to try protocol version 42, as you suggested. Renamed `cql_bad_protocol` to `cql_with_protocol`.

@nuivall I find this indirection a short, readable, and handy wrapper which also makes sure that successful session is shut down and I have introduced a successful connection before a failed one in the tests.

nuivall commented on test/cqlpy/test_protocol_exceptions.py:
```diff
@@ -0,0 +1,156 @@
+            resp = s.recv(4096)
+            assert resp[4] == 0x10, f"expected AUTH_SUCCESS, got {hex(resp[4])}"
+
+        # 4) If requested, send malformed BATCH
```
what 4) means? I don't see earlier numbers anywhere

nuivall commented:
Shouldn't you check in your tests also for "cpp_exceptions" metric to confirm that the patch works as intended?

nuivall commented on transport/server.cc:
```diff
@@ -378,23 +379,25 @@ cql_server::connection::read_frame() {
                 full.get_write()[0] = _version;
                 std::copy(tail.get(), tail.get() + tail.size(), full.get_write() + 1);
-                auto frame = parse_frame(std::move(full));
+                auto frame_res = parse_frame(std::move(full));
```
nit: not sure if it's worth to rename frame variable. especially reading this code _res suffix means nothing to me.

DarioMirovic commented on test/cqlpy/test_protocol_exceptions.py:
```diff
@@ -0,0 +1,156 @@
+@contextmanager
+def cql_bad_protocol(host_str, port, creds):
+    try:
+        # Use the default superuser credentials, which work for both Scylla and Cassandra
```
Moved the comment to the place where I'm setting the credentials

DarioMirovic commented on transport/server.cc:
```diff
@@ -675,8 +682,19 @@ thread_local cql_server::connection::execution_stage_type
+            try { ++_server._stats.errors[ex->code()]; } catch(...) {}
+            write_response(make_error(0, ex->code(), ex->what(), tracing::trace_state_ptr()));
+        } else {
+            std::rethrow_exception(eptr);
```
This is not introducing a new metric, but rather making some exceptions processed without rethrowing them via `f.get()` call.

`f.get()` throws if `f` is not computed successfully, and even if we were to remove all the throws deeper in the code, an exception would be thrown here, and caught immediately below. By checking `f.failed()` and then `f.get_exception()`, we avoid throwing the exception.

For now I only extracted `protocol_exception` to be handled differently. I believe we should have `try_catch` without rethrowing for the other exception types too, but I did not want to mix it in this PR. That's why if it's not `protocol_exception` in the `try` block, we do `std::rethrow_exception(eptr)`.

`protocol_exception` derives from `cassandra_exception`. `protocol_exception` now reports `protocol error:` , and not `connection error:` in the debug log code, but the error metrics themselves, which operate as `map[ex code] = ex message` are the same as before. So this is not a new metric.

DarioMirovic commented:
> Shouldn't you check in your tests also for "cpp_exceptions" metric to confirm that the patch works as intended?

@nuivall I have checked it already, and for `test_protocol_version_mismatch` it is 457 -> 463 with `protocol_exception` throwing, and 451 -> 453 with `protocol_exception` returning.

However, I don't see what I should assert. Even asserting that the number of cpp exceptions increased is not logically correct. This PR obviously reduced number of exceptions that happened in the full flow. It might happen in the future that there is not a single exception during this complete test.

avikivity commented:
>   
>   * Now there are 0 exceptions thrown in the `test_protocol_version_mismatch` test


Ah, good.

avikivity commented on test/cqlpy/test_protocol_exceptions.py:
```diff
@@ -0,0 +1,156 @@
+    return result
+
+@contextmanager
+def cql_bad_protocol(host_str, port, creds):
```
That number caused more damage to programmer creativity than any LLM has the hope of ever achieving.

avikivity commented:
smaller change, and isolated from the rest of the code.
> 
> As for `cpp_exceptions` - I could count the number of exceptions thrown in the test, which is less than before per last measurements, and assert that it does not increase by more than that. I am worried that, if we execute multiple tests in parallel on the same instantiated cluster, there'll be possibility for race condition and the assert for max count of exceptions will fail.
> 
> @avikivity do you have some suggestions?


I think each cluster is assigned to at most one test at a given point in time.

There could be residual exceptions, but it's easy to ignore those.

DarioMirovic commented:
> I think each cluster is assigned to at most one test at a given point in time.
> 
> There could be residual exceptions, but it's easy to ignore those.

This is good, thanks.

Then I'll just do as you suggested, run it 100 times, have some epsilon for exceptions that is visibly lower than 100 and that should do the trick.

DarioMirovic commented:
In V7:
- Removed the final exception throwing for the remaining tests
- Added zero CPP exception throws in the tests
    - As of now, there are 0 throws in `test_protocol_exceptions.py` tests

avikivity commented on transport/server.cc:
```diff
@@ -1063,8 +1078,21 @@ cql_server::connection::process(uint16_t stream, request_reader in, service::cli
+                return msg_or_error_t{ std::in_place_index<1>, f.get_exception()};
+            }
+            return msg_or_error_t{ std::in_place_index<0>, std::move(f.get()) };
+        });
```
And don't introduce new continuations to hot paths like that.

DarioMirovic replied:
Reimplemented using `co_await coroutine::as_future` and removed the `then_wrapped` continuation introduced in the previous iteration.

DarioMirovic commented:
In V8:
- Used `co_await coroutine::as_future` to avoid exception throwing instead of `then_wrapped` continuation
- Edited the test commits so that the first test commit works by itself, and the second test commit comes after the code changes that remove exceptions, making sure no commit is breaking the functionality
- Updated tests to run 100 times, with exception threshold set to 10

DarioMirovic commented on test/cqlpy/test_protocol_exceptions.py:
```diff
@@ -0,0 +1,201 @@
+
+    cpp_exception_metrics_after = get_cpp_exceptions_metrics(host)
+    assert cpp_exception_metrics_after == cpp_exception_metrics_before, "Expected C++ protocol errors to not increase after the test"
+
```
True. I fixed it - the first test commit works by itself, the second one with exceptions assertion comes after the C++ code changes.

avikivity commented:
@scylladb/releng CI says this is good, but did not run any tests

yaronkaikov commented:
> @scylladb/releng CI says this is good, but did not run any tests

CI is still running (or actually waiting for local builders to start compiling) - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18349/console

btw, you can click on one of the CI stages (Checkout or Framework test) and you would see it :-)
